crates/hir-ty/src/infer/expr.rs

```diff
@@ -933,8 +933,23 @@ impl InferenceContext<'_> {
         let prev_ret_coercion =
             mem::replace(&mut self.return_coercion, Some(CoerceMany::new(ret_ty.clone())));
 
+        let expected = &Expectation::has_type(ret_ty);
         let (_, inner_ty) = self.with_breakable_ctx(BreakableKind::Border, None, None, |this| {
-            this.infer_block(tgt_expr, *id, statements, *tail, None, &Expectation::has_type(ret_ty))
+            let ty = this.infer_block(tgt_expr, *id, statements, *tail, None, expected);
```

Comment:
I think it suffices to merely replace this by `self.infer_return(tgt_expr)`?

Reply:
I do not fully understand these functions and may be misunderstanding, but I think that we cannot use this way. First, `infer_return` method only coerces and does not return a value. Next, the method internally calls `infer_expr_inner`, which internally calls `infer_async_block`. This results in an infinite loop.

Reply:
Right